vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -321,15 +339,91 @@ and
 
     let optionsAssociation = ConditionalWeakTable<IWorkspaceProjectContext, string[]>()
 
+    member private this.ProvideProjectSiteProvider(workspace:Workspace, project:Project) =
+        let visualStudioWorkspace = workspace :?> VisualStudioWorkspace
+        let hier = visualStudioWorkspace.GetHierarchy(project.Id)
+
+        {new IProvideProjectSite with
+            member iProvideProjectSite.GetProjectSite() =
+                let sources () = 
+                    let items,_,_ = projectInfoManager.GetProjectInfo(project.FilePath)
+                    items
+                let references () =
+                    let _,references,_ = projectInfoManager.GetProjectInfo(project.FilePath)
+                    references
+                let options () =
+                    let _,references,options = projectInfoManager.GetProjectInfo(project.FilePath)
+                    Array.concat [options; references |> Array.map(fun r -> "-r:" + r)]
+                let caption () = project.Name
+                let projFileName () = project.FilePath
+                let targetFrameworkMoniker = ""
+                let projectGuid () = project.Id.Id.ToString()
+                let creationTime = System.DateTime.Now
+                let mutable errorReporter = 
+                    let reporter = ProjectExternalErrorReporter(project.Id, "FS", this.SystemServiceProvider)
+                    Some(reporter:> Microsoft.VisualStudio.Shell.Interop.IVsLanguageServiceBuildErrorReporter2)
+
+                {new Microsoft.VisualStudio.FSharp.LanguageService.IProjectSite with
+                    member __.SourceFilesOnDisk() = sources ()
+                    member __.DescriptionOfProject() = caption ()
+                    member __.CompilerFlags() = options ()
+                    member __.ProjectFileName() = projFileName ()
+                    member __.AdviseProjectSiteChanges(_,_) = ()
+                    member __.AdviseProjectSiteCleaned(_,_) = ()
+                    member __.AdviseProjectSiteClosed(_,_) = ()
+                    member __.IsIncompleteTypeCheckEnvironment = false
+                    member __.TargetFrameworkMoniker = targetFrameworkMoniker
+                    member __.ProjectGuid = projectGuid ()
+                    member __.LoadTime = creationTime
+                    member __.ProjectProvider = Some iProvideProjectSite
+                    member __.AssemblyReferences() = references ()
+                    member __.BuildErrorReporter with get () = errorReporter and 
+                                                      set (v) = errorReporter <- v
+                }
+
+        interface IVsHierarchy with
```

Comment:
I think I mentioned this in a previous review, but please document why ``IVsHierarchy`` is required here. I think it is because of whackiness in ProjectSitesAndFiles.fs where we query for this interface.